src/multi.c

```diff
@@ -382,21 +382,17 @@ int isWatchedKeyExpired(client *c) {
 /* "Touch" a key, so that if this key is being WATCHed by some client the
  * next EXEC will fail. */
 void touchWatchedKey(redisDb *db, robj *key) {
-    list *clients;
-    listIter li;
-    listNode *ln;
+    list *clients = NULL;
 
     if (dictSize(db->watched_keys) == 0) return;
-    clients = dictFetchValue(db->watched_keys, key);
-    if (!clients) return;
 
     /* Mark all the clients watching this key as CLIENT_DIRTY_CAS */
     /* Check if we are already watching for this key */
-    listRewind(clients,&li);
-    while((ln = listNext(&li))) {
-        client *c = listNodeValue(ln);
-
+    while ((clients = dictFetchValue(db->watched_keys, key))) {
+        client *c = listNodeValue(listFirst(clients));
         c->flags |= CLIENT_DIRTY_CAS;
+        /* As the client is marked as dirty, there is no point in watching other key events. */
+        unwatchAllKeys(c);
```

Comment:
arguably, the same can be done in `touchAllWatchedKeysInDb`, i.e. if we don't clean the client watched keys, they'll remain there until exec cleans them.

Reply:
> i wonder what's the implications of this "bug" / improvement? other than a minor performance and memory overhead of the old code, what else was wrong? (unless maybe some use cases watch thousands of keys?)

I didn't see a case where this caused a real issue.
While I agree that this isn't a bug, reading the code made me puzzled as to why we keep watching after getting dirty. I think this implementation is more accurate and we may gain performance in some cases as well. Anyway, I fully understand if you think this change isn't significant to be merged.

Reply:
> arguably, the same can be done in `touchAllWatchedKeysInDb`, i.e. if we don't clean the client watched keys, they'll remain there until exec cleans them.

Added